File: backend/src/main/java/de/garrafao/phitag/domain/judgement/userankrelativejudgement/UseRankRelativeJudgement.java
package de.garrafao.phitag.domain.judgement.userankrelativejudgement;

import de.garrafao.phitag.domain.annotator.Annotator;
import de.garrafao.phitag.domain.instance.userankrelative.UseRankRelativeInstance;
import de.garrafao.phitag.domain.judgement.IJudgement;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang3.Validate;

import javax.persistence.*;

@Entity
@Table(name = "phitaguserankrelativejudgement")
@Getter
public class UseRankRelativeJudgement implements IJudgement {

    @Setter
    @EmbeddedId
    private UseRankRelativeJudgementId id;

    @MapsId("annotatorid")
    @ManyToOne
    private Annotator annotator;

    @MapsId("instanceid")
    @ManyToOne
    private UseRankRelativeInstance instance;

    @Setter
    @Column(name = "rank", nullable = false)
    private String  label;

    @Setter
    @Column(name = "comment", nullable = false)
    private String comment;

    UseRankRelativeJudgement() {
    }

    public UseRankRelativeJudgement(UseRankRelativeInstance useRankRelativeInstance, Annotator annotator, String label, String comment) {
        Validate.notNull(useRankRelativeInstance, "instanceData must not be null");
        Validate.notNull(annotator, "annotator must not be null");
        Validate.notNull(label, "label must not be null");
        Validate.notNull(comment, "comment must not be null");

        this.id = new UseRankRelativeJudgementId(useRankRelativeInstance.getId(), annotator.getId());

        this.instance = useRankRelativeInstance;
        this.annotator = annotator;

        this.label = label;
        this.comment = comment;
    }

    public UseRankRelativeInstance getUseRankRelativeInstance() {
        return instance;
    }

    @Override
    public String toString() {
        return "UseRankRelativeJudgement [id=" + id + ", label=" + label + ", comment=" + comment + "]";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        UseRankRelativeJudgement other = (UseRankRelativeJudgement) obj;
        return this.id.equals(other.getId());
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}

File: frontend/lib/model/judgement/userankrelativejudgement/model/UseRankRelativeJudgement.ts
import UseRankRelativeInstance from "../../../instance/userankreltiveinstance/model/UseRankRelativeInstance";
import { IJudgement, IJudgementConstructor } from "../../model/IJudgement";
import UseRankRelativeJudgementDto from "../dto/UseRankRelativeJudgementDto";
import UseRankRelativeJudgementId from "./UseRankRelativeJudgementId";

export default class UseRankRelativeJudgement implements IJudgement {

    readonly id: UseRankRelativeJudgementId;

    readonly instance: UseRankRelativeInstance;
    
    readonly label: string;
    readonly comment: string;

    constructor(id: UseRankRelativeJudgementId, instance: UseRankRelativeInstance, label: string, comment: string) {
        this.id = id;
        this.instance = instance;
        this.label = label;
        this.comment = comment;
    }

    public getId(): UseRankRelativeJudgementId {
        return this.id;
    }

    public getInstance(): UseRankRelativeInstance {
        return this.instance;
    }

    public getLabel(): string {
        return this.label;
    }

    public getComment(): string {
        return this.comment;
    }

    public static fromDto(dto: UseRankRelativeJudgementDto): UseRankRelativeJudgement {
        return new UseRankRelativeJudgement(
            UseRankRelativeJudgementId.fromDto(dto.id),
            UseRankRelativeInstance.fromDto(dto.instance),
            dto.label,
            dto.comment
        );
    }

}

export class UseRankRelativeJudgementConstructor implements IJudgementConstructor {
    fromDto(dto: UseRankRelativeJudgementDto): UseRankRelativeJudgement {
        return UseRankRelativeJudgement.fromDto(dto);
    }
}